### ISSUE TITLE
tfe_notification_configuration creates multiple configurations

### ISSUE THREAD

**Issue by sr-mpamera** [hashicorp/terraform-provider-tfe#306]
I am terraforming notifications in tfe. The plan says it is going to create 1 notification configuration for each workspace but it creates multiple when applied.

```
Error: Error creating notification configuration slack-notification: bad request

Workspace has reached its limit of 20 Notification Configurations
```

**Comment by radditude** [comment]
Hmm, I can't seem to reproduce your issue! I'm seeing only one notification configuration per Terraform resource, as I would expect.

Could you share a bit more about your configuration and what versions of TFE and this provider you're using?

**Comment by sr-mpamera** [comment]
```
resource "tfe_notification_configuration" "slack" {
  for_each              = data.tfe_workspace_ids.all.ids
  name                   = "slack-notification"
  enabled               = true
  destination_type = "slack"
  triggers                = ["run:errored", "run:needs_attention"]
  url                         = "<slack-channel-url>"
  workspace_id      = each.value
}
```
**Plan result:**

Plan: 11 to add, 0 to change, 0 to destroy. 

This is correct as we have 11 workspaces.

**Apply result** 

tfe_notification_configuration.slack["datadog"]: Still creating... [2m50s elapsed]
tfe_notification_configuration.slack["github"]: Still creating... [2m50s elapsed]
tfe_notification_configuration.slack["pagerduty"]: Still creating... [10s elapsed]
tfe_notification_configuration.slack["pagerduty"]: Still creating... [20s elapsed]
tfe_notification_configuration.slack["pagerduty"]: Still creating... [30s elapsed]
tfe_notification_configuration.slack["pagerduty"]: Still creating... [40s elapsed]
tfe_notification_configuration.slack["pagerduty"]: Still creating... [50s elapsed]
tfe_notification_configuration.slack["pagerduty"]: Still creating... [1m0s elapsed]
tfe_notification_configuration.slack["pagerduty"]: Still creating... [1m10s elapsed]
tfe_notification_configuration.slack["pagerduty"]: Still creating... [1m20s elapsed]
tfe_notification_configuration.slack["pagerduty"]: Still creating... [1m30s elapsed]
tfe_notification_configuration.slack["pagerduty"]: Still creating... [1m40s elapsed]
tfe_notification_configuration.slack["pagerduty"]: Still creating... [1m50s elapsed]
tfe_notification_configuration.slack["pagerduty"]: Still creating... [2m0s elapsed]
tfe_notification_configuration.slack["pagerduty"]: Still creating... [2m10s elapsed]
tfe_notification_configuration.slack["pagerduty"]: Still creating... [2m20s elapsed]
tfe_notification_configuration.slack["pagerduty"]: Still creating... [2m30s elapsed]

Error: Error creating notification configuration slack-notification: bad request

Workspace has reached its limit of 20 Notification Configurations

**Comment by radditude** [comment]
Interesting, I'm not seeing the same problem, even with that exact configuration! I'm sure you've checked this already, but could you confirm that those workspaces didn't have existing notification configurations set up before you ran Terraform? As the error message says, there's a limit of 20 per workspace.

Failing that, I'll need more to go on to diagnose the issue - could you set [the `TF_LOG` environment variable](https://www.terraform.io/docs/internals/debugging.html) to `DEBUG` and paste the output here?

**Comment by sr-mpamera** [comment]
@radditude I have set the TF_LOG env variables and got this output. 

```
2021/05/10 16:30:45 [INFO] Terraform version: 0.14.8  
2021/05/10 16:30:45 [INFO] Go runtime version: go1.16
2021/05/10 16:30:45 [INFO] CLI args: []string{"/usr/local/bin/terraform", "init"}
2021/05/10 16:30:45 [DEBUG] Attempting to open CLI config file: /Users/mamathapamera/.terraformrc
2021/05/10 16:30:45 Loading CLI configuration from /Users/mamathapamera/.terraformrc
2021/05/10 16:30:45 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2021/05/10 16:30:45 [DEBUG] ignoring non-existing provider search directory /Users/mamathapamera/.terraform.d/plugins
2021/05/10 16:30:45 [DEBUG] ignoring non-existing provider search directory /Users/mamathapamera/Library/Application Support/io.terraform/plugins
2021/05/10 16:30:45 [DEBUG] ignoring non-existing provider search directory /Library/Application Support/io.terraform/plugins
2021/05/10 16:30:45 [INFO] CLI command args: []string{"init"}
2021/05/10 16:30:45 [WARN] Log levels other than TRACE are currently unreliable, and are supported only for backward compatibility.
  Use TF_LOG=TRACE to see Terraform's internal logs.
  ----
2021/05/10 16:30:45 [DEBUG] Service discovery for app.terraform.io at https://app.terraform.io/.well-known/terraform.json
2021/05/10 16:30:45 [DEBUG] Retrieve version constraints for service tfe.v2.1 and product terraform
2021/05/10 16:30:46 [DEBUG] checking for provisioner in "."
2021/05/10 16:30:46 [DEBUG] checking for provisioner in "/usr/local/bin"
2021/05/10 16:30:46 [INFO] Failed to read plugin lock file .terraform/plugins/darwin_amd64/lock.json: open .terraform/plugins/darwin_amd64/lock.json: no such file or directory
2021/05/10 16:30:46 [DEBUG] Service discovery for registry.terraform.io at https://registry.terraform.io/.well-known/terraform.json
2021/05/10 16:30:47 [WARN] Log levels other than TRACE are currently unreliable, and are supported only for backward compatibility.
  Use TF_LOG=TRACE to see Terraform's internal logs.
  ----
2021/05/10 16:30:47 [DEBUG] GET https://registry.terraform.io/v1/providers/hashicorp/tfe/versions
```

**Comment by radditude** [comment]
Ah, I should have specified - I need the debug logs from a `terraform apply` operation where you're seeing the `Workspace has reached its limit of 20 Notification Configurations` error. Since I can't reproduce the problem myself, I'm hoping to see the API calls the resource is making during the apply so I can identify where the extra notification configurations might be coming from.

**Comment by sr-mpamera** [comment]
@radditude How do I turn on the debug logs for terraform apply in TFE?

**Comment by radditude** [comment]
You'll need to add the `TF_LOG` environment variable under the workspace's variable settings and set it to `DEBUG`, then queue a new run! You can download the full log as a text file using the "view raw log" button in the run view, which is a bit nicer than trying to copy and paste from the web terminal:

<img width="325" alt="Screen Shot 2021-05-17 at 1 48 34 PM" src="https://user-images.githubusercontent.com/17039873/118554992-015dfe00-b717-11eb-8709-12c44edfd2c8.png">

**Comment by flatiron32** [comment]
Hello @radditude I am working with @sr-mpamera on this. 
The logs include the `Authorization` header so I don't want to share it here but what I see is 
```2021/05/19 02:43:51 [DEBUG] tfe_notification_configuration.slack: applying the planned Create change
2021-05-19T02:43:51.547Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: 2021/05/19 02:43:51 [DEBUG] setting computed for "email_addresses" from ComputedKeys
2021-05-19T02:43:51.547Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: 2021/05/19 02:43:51 [DEBUG] setting computed for "email_user_ids" from ComputedKeys
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: 2021/05/19 02:43:51 [DEBUG] Create notification configuration: slack-notification
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: 2021/05/19 02:43:51 [DEBUG] TFE API Request Details:
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: ---[ REQUEST ]---------------------------------------
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: POST /api/v2/workspaces/ws-h8JrZHpbK22gcT8Z/notification-configurations HTTP/1.1
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: Host: app.terraform.io
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: User-Agent: go-tfe
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: Content-Length: 266
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: Accept: application/vnd.api+json
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: Authorization: Bearer *redacted*
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: Content-Type: application/vnd.api+json
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: Accept-Encoding: gzip
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: 
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: {
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:  "data": {
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:   "type": "notification-configurations",
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:   "attributes": {
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:    "destination-type": "slack",
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:    "enabled": true,
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:    "name": "slack-notification",
n": "",
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:    "triggers": [
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:     "run:completed",
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:     "run:needs_attention",
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:     "run:errored"
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:    ],
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:    "url": "https://app.slack.com/client/T19855G21/CMRPM9Q0Y"
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:   }
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:  }
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: }
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: 
2021-05-19T02:43:51.548Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: -----------------------------------------------------
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: 2021/05/19 02:43:51 [DEBUG] TFE API Response Details:
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: ---[ RESPONSE ]--------------------------------------
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: HTTP/1.1 500 Internal Server Error
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: Content-Length: 61
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: Cache-Control: no-cache
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: Connection: keep-alive
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: Content-Type: application/vnd.api+json; charset=utf-8
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: Date: Wed, 19 May 2021 02:43:51 GMT
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: Referrer-Policy: strict-origin-when-cross-origin
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: Tfp-Api-Version: 2.3
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: Vary: Accept-Encoding
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: Vary: Origin
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: X-Content-Type-Options: nosniff
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: X-Download-Options: noopen
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: X-Frame-Options: SAMEORIGIN
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: X-Permitted-Cross-Domain-Policies: none
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: X-Ratelimit-Limit: 30
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: X-Ratelimit-Remaining: 27
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: X-Ratelimit-Reset: 0.34
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: X-Request-Id: 998d4685-16a0-5c58-3a6e-ce2f293e6f13
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: X-Xss-Protection: 1; mode=block
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: 
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: {
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:  "errors": [
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:   {
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:    "status": "500",
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:    "title": "internal server error"
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:   }
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5:  ]
2021-05-19T02:43:51.666Z [DEBUG] plugin.terraform-provider-tfe_v0.24.0_x5: }
```

There are 20 of these POSTs that fail with 500. But then the 21st one fails with `400 Bad Request` because out account has used all 20 of the notifications allotted. 

It looks like the provider of GO client is not handling a 5XX response how I would expect. My expectation is the apply would halt when an internal server error is encountered.

I am able to recreate this issue by POSTing this body to the API. I suspect the problem is that we are not providing a webhook URL, but a URL for a channel. If I change it to a webhook URL it work. I would guess the API is not handling data validation and failure to connect the URL very well and this should be a 4XX error wit information to the client.

In summary, I think this is more a bug with Terraform Cloud. Can you point me to where I should open a bug against that product?

**Comment by radditude** [comment]
Ah, I see! I think retrying 500 errors is intentional, actually, to handle the case where a server is experiencing some small blips in availability (say, a Terraform Enterprise instance with a misconfigured load balancer) - in that case, the error is transient and may resolve with a retry, so it shouldn't necessarily stop the apply cold.

As I see it, the bug is that the notification configuration gets created at all. If there's a validation error, we should return the error and _not_ create the thing, so that Terraform (or some other API client) can safely retry the call.

If you're a TFE or TFC customer, submitting a ticket to our support team would be very helpful in making sure this gets triaged and prioritized. Thank you!
